confd/warp/warp.go

```diff
@@ -0,0 +1,228 @@
+	}
+	return categories[i].Order > categories[j].Order
+}
+
+func (categories Categories) Swap(i, j int) {
+	categories[i], categories[j] = categories[j], categories[i]
+}
+
+// Entry link in the warp menu
+type Entry struct {
+	DisplayName string
+	Href        string
+	Title       string
+	Target      string
+}
+
+// Entries is a collection of warp entries
+type Entries []Entry
+
+// sort methods
+
+func (entries Entries) Len() int {
+	return len(entries)
+}
+
+func (entries Entries) Less(i, j int) bool {
+	return entries[i].DisplayName < entries[j].DisplayName
+}
+
+func (entries Entries) Swap(i, j int) {
+	entries[i], entries[j] = entries[j], entries[i]
+}
+
+func isKeyNotFound(err error) bool {
+	if cErr, ok := err.(client.Error); ok {
+		return cErr.Code == client.ErrorCodeKeyNotFound
+	}
+	return false
+}
+
+func unmarshalDogu(kapi client.KeysAPI, key string) (RawData, error) {
+	resp, err := kapi.Get(context.Background(), key+"/current", nil)
+	if err != nil {
+		// the dogu seems to be unregistered
+		if isKeyNotFound(err) {
+			return nil, nil
+		}
+		return nil, errors.Wrapf(err, "failed to read key %s from etcd", key)
+	}
+
+	version := resp.Node.Value
+	resp, err = kapi.Get(context.Background(), key+"/"+version, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read version child from key %s", key)
+	}
+
+	dogu := RawData{}
+	err = json.Unmarshal([]byte(resp.Node.Value), &dogu)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshall json from etcd")
+	}
+
+	return dogu, nil
+}
+
+func unmarshalExternal(kapi client.KeysAPI, key string) (RawData, error) {
+	resp, err := kapi.Get(context.Background(), key, nil)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to read key %s from etcd", key)
+	}
+	external := RawData{}
+	err = json.Unmarshal([]byte(resp.Node.Value), &external)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to unmarshall json from etcd")
+	}
+
+	return external, nil
+}
+
+func createHref(dogu RawData) string {
+	// remove namespace
+	parts := strings.Split(dogu["Name"].(string), "/")
+	return "/" + parts[len(parts)-1]
+}
+
+func createDoguEntry(element RawData) Entry {
+	return Entry{
+		DisplayName: element["DisplayName"].(string),
+		Href:        createHref(element),
+		Title:       element["Description"].(string),
+		Target:      "self",
+	}
+}
+
+func createExternalEntry(element RawData) Entry {
+	return Entry{
+		DisplayName: element["DisplayName"].(string),
+		Href:        element["URL"].(string),
+		Title:       element["Description"].(string),
+		Target:      "external",
+	}
+}
+
+func filterByTag(dogus []RawData, tag string) []RawData {
+	filtered := []RawData{}
+	for _, raw := range dogus {
+		if raw["Tags"] != nil {
+			tags := raw["Tags"].([]interface{})
+			if tags != nil && Contains(tags, tag) {
+				filtered = append(filtered, raw)
+			}
+		}
+	}
+	return filtered
+}
+
+func (destination *Categories) insertCategories(newCategories Categories) {
```

Comment:
**[checkstyle]** <sub>reported by [reviewdog](https://github.com/haya14busa/reviewdog) :dog:</sub>
receiver name destination should be consistent with previous receiver name categories for Categories